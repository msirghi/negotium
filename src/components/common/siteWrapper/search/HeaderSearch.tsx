import { TextField } from '@mui/material';
import { makeStyles } from '@mui/styles';
import colors from '../../../../common/styles/colors';
import Autocomplete from '@mui/material/Autocomplete';
import { useSelector } from 'react-redux';
import { RootState } from '../../../../redux/store';
import { useEffect, useState } from 'react';
import { HeaderSearchOptions } from './types';
import { useRouter } from 'next/router';
import ReorderIcon from '@mui/icons-material/Reorder';
import { useIsMobile } from '../../../../common/hooks/common/useIsMobile';
import { HeaderSearchOption } from './option/HeaderSearchOption';

const useStyles = makeStyles({
  input: {
    backgroundColor: colors.greys['800'],
    borderRadius: 5,
    marginLeft: 20,
    transition: '.3s all ease-in-out',
    '&.Mui-focused': {
      backgroundColor: colors.white,
      border: 'none',
      width: 400,
    },
  },
});

export const HeaderSearch = () => {
  const classes = useStyles();
  const [options, setOptions] = useState<HeaderSearchOptions[]>([]);
  const projects = useSelector((state: RootState) => state.projects.projects);
  const router = useRouter();
  const [open, setOpen] = useState(false);
  const isMobile = useIsMobile();
  const [selectedOption, setSelectedOption] = useState<string | null>(null);

  const closePopper = () => {
    setOpen(false);
  };

  const openPopper = () => {
    setTimeout(() => {
      setOpen(true);
    }, 300);
  };

  const configureProjects = () => {
    const projectOptions: HeaderSearchOptions[] = projects.map((project) => {
      return {
        id: project.id,
        title: project.name,
        onClick: () => {
          closePopper();
          router.push(`/home/projects/${project.id}`);
          setSelectedOption(project.id);
        },
        Icon: ReorderIcon,
      };
    });
    setOptions((prevState) =>
      Array.from(new Set([...prevState, ...projectOptions]))
    );
  };

  useEffect(() => {
    if (projects) {
      configureProjects();
    }
  }, [projects]);

  if (isMobile) {
    return <div />;
  }

  return (
    <Autocomplete
      key={selectedOption}
      disableClearable
      disabledItemsFocusable
      sx={{ width: 300 }}
      onChange={(e, value) => {
        value.onClick();
      }}
      size={'small'}
      id={'header-autocomplete'}
      className={classes.input}
      disablePortal
      open={open}
      onOpen={openPopper}
      onClose={closePopper}
      getOptionLabel={() => ''}
      renderInput={(params) => (
        <TextField {...params} size={'small'} placeholder={'Search'} />
      )}
      options={Array.from(new Set(options))}
      renderOption={(props, option) => (
        <HeaderSearchOption props={props} option={option} />
      )}
    />
  );
};
