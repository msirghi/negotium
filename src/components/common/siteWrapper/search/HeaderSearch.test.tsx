import { projectsMock } from '../../../../common/tests/mockData/projects-mock';
import {
  MockReduxProvider,
  MockThemeProvider,
} from '../../../../common/tests/TestUtils';
import { HeaderSearch } from './HeaderSearch';
import renderer, { act } from 'react-test-renderer';
import { mount } from 'enzyme';

jest.mock('next/router', () => ({
  useRouter: () => ({
    route: 'inbox',
    query: {},
  }),
}));

describe('HeaderSearch', () => {
  const reduxStore = {
    projects: [...projectsMock],
  };

  const renderComponent = () => {
    return (
      <MockReduxProvider reduxStore={reduxStore}>
        <MockThemeProvider>
          <HeaderSearch />
        </MockThemeProvider>
      </MockReduxProvider>
    );
  };

  it('should match the snapshot', () => {
    const tree = renderer.create(renderComponent());
    expect(tree).toMatchSnapshot();
  });

  it('should handle onChange autocomplete event', () => {
    const wrapper = mount(renderComponent());
    const autocomplete = wrapper.find('#header-autocomplete').at(0) as any;
    const spy = jest.fn();

    act(() => {
      autocomplete.props().onChange({} as any, { onClick: spy });
    });
    expect(spy).toBeCalled();
  });

  it('should handle open state', () => {
    jest.useFakeTimers();
    const wrapper = mount(renderComponent());
    const autocomplete = wrapper.find('#header-autocomplete').at(0) as any;

    act(() => {
      autocomplete.props().onOpen();
    });
    jest.runOnlyPendingTimers();
    wrapper.update();
    expect(autocomplete.props().open).toBeDefined();
  });

  it('should handle close state', () => {
    jest.useFakeTimers();
    const wrapper = mount(renderComponent());
    const autocomplete = wrapper.find('#header-autocomplete').at(0) as any;

    act(() => {
      autocomplete.props().onClose();
    });
    jest.runOnlyPendingTimers();
    wrapper.update();
    expect(autocomplete.props().open).toBeDefined();
  });

  it('should return value for option label', () => {
    jest.useFakeTimers();
    const wrapper = mount(renderComponent());
    const autocomplete = wrapper.find('#header-autocomplete').at(0) as any;

    const result = autocomplete.props().getOptionLabel({title: 'title'});
    expect(result).toEqual('title');
  });
});
