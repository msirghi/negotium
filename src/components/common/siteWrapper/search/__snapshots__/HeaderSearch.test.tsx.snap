// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HeaderSearch should match the snapshot 1`] = `
<div
  aria-expanded={false}
  aria-owns={null}
  className="MuiAutocomplete-root MuiAutocomplete-hasPopupIcon makeStyles-input-1 makeStyles-input-2 css-dst279-MuiAutocomplete-root"
  onClick={[Function]}
  onKeyDown={[Function]}
  onMouseDown={[Function]}
  role="combobox"
>
  <div
    className="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
  >
    <div
      className="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-sizeSmall MuiInputBase-adornedEnd MuiAutocomplete-inputRoot css-brkcn7-MuiInputBase-root-MuiOutlinedInput-root"
      onClick={[Function]}
    >
      <input
        aria-activedescendant={null}
        aria-autocomplete="list"
        aria-controls={null}
        aria-invalid={false}
        autoCapitalize="none"
        autoComplete="off"
        autoFocus={false}
        className="MuiOutlinedInput-input MuiInputBase-input MuiInputBase-inputSizeSmall MuiInputBase-inputAdornedEnd MuiAutocomplete-input MuiAutocomplete-inputFocused MuiAutocomplete-input MuiAutocomplete-inputFocused css-19qh8xo-MuiInputBase-input-MuiOutlinedInput-input"
        disabled={false}
        id="header-autocomplete"
        onAnimationStart={[Function]}
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        onMouseDown={[Function]}
        placeholder="Search"
        required={false}
        spellCheck="false"
        type="text"
        value=""
      />
      <div
        className="MuiAutocomplete-endAdornment css-1q60rmi-MuiAutocomplete-endAdornment"
      >
        <button
          aria-label="Open"
          className="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium MuiAutocomplete-popupIndicator css-qzbt6i-MuiButtonBase-root-MuiIconButton-root-MuiAutocomplete-popupIndicator"
          disabled={false}
          onBlur={[Function]}
          onClick={[Function]}
          onContextMenu={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={-1}
          title="Open"
          type="button"
        >
          <svg
            aria-hidden={true}
            className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="ArrowDropDownIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M7 10l5 5 5-5z"
            />
          </svg>
        </button>
      </div>
      <fieldset
        aria-hidden={true}
        className="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
      >
        <legend
          className="css-1a1ihhj"
        >
          <span
            className="notranslate"
            dangerouslySetInnerHTML={
              Object {
                "__html": "&#8203;",
              }
            }
          />
        </legend>
      </fieldset>
    </div>
  </div>
</div>
`;
